File: src/DepositfilesSDK/DepositfilesSDK/Interfaces/RootClient.vb
Imports Newtonsoft.Json
Imports DepositfilesSDK.JSON

Public Class RootClient
    Implements IRoot




#Region "Search"
    Public Async Function _Search(SearchKeyword As String, SearchType As utilitiez.SearchTypeEnum) As Task(Of List(Of JSON_FileMetadata)) Implements IRoot.SearchFiles
        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("file/listing", New AuthDictionary)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Using content As Net.Http.HttpContent = response.Content
                    Dim result As String = Await content.ReadAsStringAsync()

                    If Linq.JObject.Parse(result)("status").ToString = "OK" Then
                        If Linq.JObject.Parse(result)("data").ToString = "[]" Then
                            Return New List(Of JSON_FileMetadata)
                        Else
                            Dim nfo = JsonConvert.DeserializeObject(Of JSON_ListFiles)(result, JSONhandler)
                            Dim lst As New List(Of JSON_FileMetadata)
                            Select Case SearchType
                                Case utilitiez.SearchTypeEnum.Contains
                                    For Each onz In nfo.FilesList
                                        If onz.Value.filename.ToLower.Contains(SearchKeyword.ToLower) Then lst.Add(onz.Value)
                                    Next
                                Case utilitiez.SearchTypeEnum.Exact
                                    For Each onz In nfo.FilesList
                                        If onz.Value.filename.ToLower = (SearchKeyword.ToLower) Then lst.Add(onz.Value)
                                    Next
                                Case utilitiez.SearchTypeEnum.StartWith
                                    For Each onz In nfo.FilesList
                                        If onz.Value.filename.ToLower.StartsWith(SearchKeyword.ToLower) Then lst.Add(onz.Value)
                                    Next
                                Case utilitiez.SearchTypeEnum.EndWith
                                    For Each onz In nfo.FilesList
                                        If onz.Value.filename.ToLower.EndsWith(SearchKeyword.ToLower) Then lst.Add(onz.Value)
                                    Next
                                Case utilitiez.SearchTypeEnum.Ext
                                    For Each onz In nfo.FilesList
                                        If IO.Path.GetExtension(onz.Value.filename).Trim(".") = (SearchKeyword).Trim(".") Then lst.Add(onz.Value)
                                    Next
                            End Select

                            Return lst

                        End If
                    Else
                        ShowError(result)
                    End If
                End Using
            End Using
        End Using
    End Function
#End Region


#Region "CreateNewFolder"
    Public Async Function _CreateNewFolder(FolderName As String) As Task(Of JSON_FolderMetadata) Implements IRoot.CreateNewFolder
        Using localHttpClient As New HttpClient(New HCHandler)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(New pUri("folder/add", New AuthDictionary() From {{"name", FolderName}})).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If result.CheckStatus Then
                    Return JsonConvert.DeserializeObject(Of JSON_FolderMetadata)(result.Jobj.SelectToken("data").ToString, JSONhandler)
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

#Region "List"
    Public Async Function GET_List() As Task(Of JSON_List) Implements IRoot.ListFiles
        Dim parameters = New AuthDictionary() From {{"folder_id", "_root"}}

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("file/listing", parameters)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If result.CheckStatus Then
                    If result.Jobj("data").ToString = "[]" Then
                        Return New JSON_List With {.FilesList = New Dictionary(Of String, JSON_FileMetadata)}
                    Else
                        Return JsonConvert.DeserializeObject(Of JSON_List)(result, JSONhandler)
                    End If
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

#Region "ListFolders"
    Public Async Function GET_ListFolders() As Task(Of List(Of JSON_FolderMetadata)) Implements IRoot.ListFolders
        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("folder/listing", New AuthDictionary)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If result.CheckStatus Then
                    If result.Jobj("data").ToString = "[]" Then
                        Return New List(Of JSON_FolderMetadata)
                    Else
                        Return JsonConvert.DeserializeObject(Of List(Of JSON_FolderMetadata))(result.Jobj.SelectToken("data").ToString, JSONhandler)
                    End If
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

End Class

File: DepositfilesSDK/DepositfilesSDK/Interfaces/RootClient.cs
using DepositfilesSDK.JSON;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Collections.Generic;
using System.Net.Http;
using System.Threading.Tasks;
using static DepositfilesSDK.Basic;
using static DepositfilesSDK.Utilitiez;

namespace DepositfilesSDK
{
    public class RootClient : IRoot
    {

        public string RootID
        {
            get
            {
                return "_root";
            }
        }

        public async Task<List<JSON_FileMetadata>> SearchFiles(string SearchKeyword, SearchTypeEnum SearchType)
        {
            using (HtpClient localHttpClient = new HtpClient(new HCHandler()))
            {
                using (HttpResponseMessage response = await localHttpClient.GetAsync(new pUri("file/listing", new AuthDictionary())).ConfigureAwait(false))
                {
                    string result = await response.Content.ReadAsStringAsync();

                    if (JObject.Parse(result).SelectToken("status").ToString() == "OK")
                    {
                        if (JObject.Parse(result).SelectToken("data").ToString() == "[]")
                        {
                            return new List<JSON_FileMetadata>();
                        }
                        else
                        {
                            var nfo = JsonConvert.DeserializeObject<JSON_ListFiles>(result, JSONhandler);
                            List<JSON_FileMetadata> lst = new List<JSON_FileMetadata>();
                            switch (SearchType)
                            {
                                case SearchTypeEnum.Contains:
                                    foreach (var onz in nfo.FilesList)
                                    {
                                        if (onz.Value.filename.ToLower().Contains(SearchKeyword.ToLower()))
                                        {
                                            lst.Add(onz.Value);
                                        }
                                    }

                                    break;
                                case SearchTypeEnum.Exact:
                                    foreach (var onz in nfo.FilesList)
                                    {
                                        if (onz.Value.filename.ToLower() == (SearchKeyword.ToLower()))
                                        {
                                            lst.Add(onz.Value);
                                        }
                                    }

                                    break;
                                case SearchTypeEnum.StartWith:
                                    foreach (var onz in nfo.FilesList)
                                    {
                                        if (onz.Value.filename.ToLower().StartsWith(SearchKeyword.ToLower()))
                                        {
                                            lst.Add(onz.Value);
                                        }
                                    }

                                    break;
                                case SearchTypeEnum.EndWith:
                                    foreach (var onz in nfo.FilesList)
                                    {
                                        if (onz.Value.filename.ToLower().EndsWith(SearchKeyword.ToLower()))
                                        {
                                            lst.Add(onz.Value);
                                        }
                                    }

                                    break;
                                case SearchTypeEnum.Ext:
                                    foreach (var onz in nfo.FilesList)
                                    {
                                        if (System.IO.Path.GetExtension(onz.Value.filename).Trim('.') == (SearchKeyword).Trim('.'))
                                        {
                                            lst.Add(onz.Value);
                                        }
                                    }

                                    break;
                            }

                            return lst;
                        }
                    }
                    else
                    {
                        throw ShowError(result);
                    }
                }
            }
        }

        public async Task<JSON_FolderMetadata> CreateNewFolder(string FolderName)
        {
            using (HtpClient localHttpClient = new HtpClient(new HCHandler()))
            {
                using (HttpResponseMessage response = await localHttpClient.GetAsync(new pUri("folder/add", new AuthDictionary() { { "name", FolderName } })).ConfigureAwait(false))
                {
                    string result = await response.Content.ReadAsStringAsync();

                    return result.CheckStatus() ? JsonConvert.DeserializeObject<JSON_FolderMetadata>(result.Jobj().SelectToken("data").ToString(), JSONhandler) : throw ShowError(result);
                }
            }
        }

        public async Task<JSON_List> ListFiles()
        {
            using (HttpClient localHttpClient = new HttpClient(new HCHandler()))
            {
                var RequestUri = new pUri("file/listing", new AuthDictionary() { { "folder_id", "_root" } });
                using (HttpResponseMessage response = await localHttpClient.GetAsync(RequestUri).ConfigureAwait(false))
                {
                    string result = await response.Content.ReadAsStringAsync();

                    if (result.CheckStatus())
                    {
                        return result.Jobj().SelectToken("data").ToString() == "[]" ? new JSON_List() { FilesList = new Dictionary<string, JSON_FileMetadata>() } : JsonConvert.DeserializeObject<JSON_List>(result, JSONhandler);
                    }
                    else
                    {
                        throw ShowError(result);
                    }
                }
            }
        }

        public async Task<List<JSON_FolderMetadata>> ListFolders()
        {
            using (HttpClient localHttpClient = new HttpClient(new HCHandler()))
            {
                using (HttpResponseMessage response = await localHttpClient.GetAsync(new pUri("folder/listing", new AuthDictionary())).ConfigureAwait(false))
                {
                    string result = await response.Content.ReadAsStringAsync();

                    if (result.CheckStatus())
                    {
                        return result.Jobj().SelectToken("data").ToString() == "[]" ? new List<JSON_FolderMetadata>() : JsonConvert.DeserializeObject<List<JSON_FolderMetadata>>(result.Jobj().SelectToken("data").ToString(), JSONhandler);
                    }
                    else
                    {
                        throw ShowError(result);
                    }
                }
            }
        }

    }
}

